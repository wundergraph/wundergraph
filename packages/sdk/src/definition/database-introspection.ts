import { IntrospectionCacheConfiguration, introspectWithCache } from './introspection-cache';
import { DatabaseSchema, mongodb, mysql, planetscale, postgresql, prisma, sqlite, sqlserver } from '../db/types';
import {
	Api,
	ApiIntrospectionOptions,
	ApiType,
	DatabaseApiCustom,
	DatabaseIntrospection,
	DataSource,
	MongoDBApi,
	MySQLApi,
	PlanetscaleApi,
	PostgresqlApi,
	PrismaApi,
	PrismaIntrospection,
	SQLiteApi,
	SQLServerApi,
} from './index';
import { introspectPrismaDatabaseWithRetries } from '../db/introspection';
import { buildSchema, parse, print } from 'graphql/index';
import { configuration } from '../graphql/configuration';
import {
	applyNameSpaceToFieldConfigurations,
	applyNameSpaceToGraphQLSchema,
	applyNameSpaceToTypeFields,
	generateTypeConfigurationsForNamespace,
} from './namespacing';
import { mapInputVariable, resolveVariable } from '../configure/variables';
import { DataSourceKind, FieldConfiguration, SingleTypeField, TypeConfiguration } from '@wundergraph/protobuf';
import { fileHash, urlHash } from '../localcache';

const introspectDatabase = async (
	introspection: DatabaseIntrospection,
	databaseSchema: DatabaseSchema,
	maxRetries: number
) => {
	const {
		success,
		message,
		graphql_schema,
		prisma_schema,
		interpolateVariableDefinitionAsJSON,
		jsonTypeFields,
		jsonResponseFields,
	} = await introspectPrismaDatabaseWithRetries(introspection, databaseSchema, maxRetries);
	if (!success) {
		return Promise.reject(message);
	}
	const schemaDocumentNode = parse(graphql_schema);
	const schema = print(schemaDocumentNode);
	const { RootNodes, ChildNodes, Fields } = configuration(schemaDocumentNode, {
		url: '',
		schemaExtension: introspection.schemaExtension,
	});
	const jsonFields = [...jsonTypeFields, ...jsonResponseFields];
	jsonFields.forEach((field) => {
		const fieldConfig = Fields.find((f) => f.typeName == field.typeName && f.fieldName == field.fieldName);
		if (fieldConfig) {
			fieldConfig.unescapeResponseJson = true;
		} else {
			Fields.push({
				fieldName: field.fieldName,
				typeName: field.typeName,
				unescapeResponseJson: true,
				argumentsConfiguration: [],
				path: [],
				requiresFields: [],
				disableDefaultFieldMapping: false,
			});
		}
	});
	const graphQLSchema = buildSchema(schema);
	const dataSource: DataSource<DatabaseApiCustom> = {
		Kind: databaseSchemaToKind(databaseSchema),
		RootNodes: applyNameSpaceToTypeFields(RootNodes, graphQLSchema, introspection.apiNamespace),
		ChildNodes: applyNameSpaceToTypeFields(ChildNodes, graphQLSchema, introspection.apiNamespace),
		Custom: {
			prisma_schema: prisma_schema,
			databaseURL: mapInputVariable(introspection.databaseURL),
			graphql_schema: schema,
			jsonTypeFields: applyNameSpaceToSingleTypeFields(jsonTypeFields, introspection.apiNamespace),
			jsonInputVariables: applyNameSpaceToTypeNames(interpolateVariableDefinitionAsJSON, introspection.apiNamespace),
		},
		Directives: [],
		RequestTimeoutSeconds: introspection.requestTimeoutSeconds ?? 0,
	};
	const dataSources: DataSource<DatabaseApiCustom>[] = [];
	dataSource.RootNodes.forEach((rootNode) => {
		rootNode.fieldNames.forEach((field) => {
			dataSources.push({
				...Object.assign({}, dataSource),
				RootNodes: [
					{
						typeName: rootNode.typeName,
						fieldNames: [field],
					},
				],
			});
		});
	});
	return {
		schema: applyNameSpaceToGraphQLSchema(schema, [], introspection.apiNamespace),
		dataSources: dataSources,
		fields: applyNameSpaceToFieldConfigurations(Fields, graphQLSchema, [], introspection.apiNamespace),
		types: generateTypeConfigurationsForNamespace(schema, introspection.apiNamespace),
		interpolateVariableDefinitionAsJSON: applyNameSpaceToTypeNames(
			interpolateVariableDefinitionAsJSON,
			introspection.apiNamespace
		),
	};
};

const databaseSchemaToKind = (schema: DatabaseSchema): DataSourceKind => {
	switch (schema) {
		case planetscale:
			return DataSourceKind.MYSQL;
		case mysql:
			return DataSourceKind.MYSQL;
		case postgresql:
			return DataSourceKind.POSTGRESQL;
		case sqlite:
			return DataSourceKind.SQLITE;
		case sqlserver:
			return DataSourceKind.SQLSERVER;
		case mongodb:
			return DataSourceKind.MONGODB;
		case prisma:
			return DataSourceKind.PRISMA;
		default:
			throw new Error(`databaseSchemaToKind not implemented for: ${schema}`);
	}
};

const applyNameSpaceToSingleTypeFields = (typeFields: SingleTypeField[], namespace?: string): SingleTypeField[] => {
	if (!namespace) {
		return typeFields;
	}
	return typeFields.map((typeField) => ({
		...typeField,
		typeName: `${namespace}_${typeField.typeName}`,
	}));
};

const applyNameSpaceToTypeNames = (typeNames: string[], namespace?: string): string[] => {
	if (!namespace) {
		return typeNames;
	}
	return typeNames.map((typeName) => {
		return `${namespace}_${typeName}`;
	});
};

type databaseConstructor<T> = (
	schema: string,
	namespace: string,
	dataSources: DataSource<DatabaseApiCustom>[],
	fields: FieldConfiguration[],
	types: TypeConfiguration[],
	interpolateVariableDefinitionAsJSON: string[],
	customJsonScalars?: string[] | undefined
) => T;

type DbApi = Api<DatabaseApiCustom>;

function introspectDatabaseWithCache<TApi extends Api<ApiType>>(
	ctor: databaseConstructor<TApi>,
	databaseUrlSchema: DatabaseSchema
) {
	const generator = async (introspection: DatabaseIntrospection, _: ApiIntrospectionOptions) => {
		const { schema, fields, types, dataSources, interpolateVariableDefinitionAsJSON } = await introspectDatabase(
			introspection,
			databaseUrlSchema,
			5
		);
		return ctor(
			schema,
			introspection.apiNamespace || '',
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON
		);
	};

	return async (introspection: DatabaseIntrospection) => {
		const url = resolveVariable(introspection.databaseURL);
		const hash = await urlHash(url);
		// Introspecting a database might be expensive, even if it's a local network URL.
		// Consider all sources as remote to cache them more aggressively
		const cacheConfig = { keyInput: hash };
		return introspectWithCache(introspection, cacheConfig, generator);
	};
}

export const introspectPostgresql = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new PostgresqlApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	postgresql
);
export const introspectMySQL = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new MySQLApi(schema, namespace, dataSources, fields, types, interpolateVariableDefinitionAsJSON, customJsonScalars),
	mysql
);
export const introspectPlanetScale = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new PlanetscaleApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	planetscale
);
export const introspectSQLite = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new SQLiteApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	sqlite
);
export const introspectSQLServer = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new SQLServerApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	sqlserver
);
export const introspectMongoDB = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new MongoDBApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	mongodb
);

export const introspectPrisma = async (introspection: PrismaIntrospection) => {
	const cacheConfig: IntrospectionCacheConfiguration = {
		keyInput: await fileHash(introspection.prismaFilePath),
		dataSource: 'localFilesystem',
	};
	return introspectWithCache(
		introspection,
		cacheConfig,
		async (introspection: PrismaIntrospection, _: ApiIntrospectionOptions): Promise<PrismaApi> => {
			const { schema, fields, types, dataSources, interpolateVariableDefinitionAsJSON } = await introspectDatabase(
				{ ...introspection, databaseURL: introspection.prismaFilePath },
				prisma,
				5
			);
			return new PrismaApi(
				schema,
				introspection.apiNamespace || '',
				dataSources,
				fields,
				types,
				interpolateVariableDefinitionAsJSON
			);
		}
	);
};
