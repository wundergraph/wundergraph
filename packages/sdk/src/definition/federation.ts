import { GraphQLSchema } from 'graphql';
import axios from 'axios';

export const isFederationService = (schema: GraphQLSchema): boolean => {
	const queryType = schema.getQueryType();
	if (queryType === undefined || queryType === null) {
		return false;
	}
	const fields = queryType.getFields();
	if (fields === undefined) {
		return false;
	}
	return Object.keys(fields).indexOf('_service') !== -1;
};

export const fetchFederationServiceSDL = async (url: string): Promise<string> => {
	const data = JSON.stringify({
		query: '{_service{sdl}}',
	});
	const res = await axios.post(url, data, {
		headers: {
			'Content-Type': 'application/json',
			Accept: 'application/json',
		},
	});
	return res.data.data._service.sdl;
};
