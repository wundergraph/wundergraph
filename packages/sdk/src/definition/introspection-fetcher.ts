import axiosRetry from 'axios-retry';
import axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from 'axios';

let axiosInstance: AxiosInstance | undefined;

export const Fetcher = (): AxiosInstance => {
	if (axiosInstance) {
		return axiosInstance;
	}

	axiosInstance = initAxios();
	return axiosInstance;
};

const initAxios = (): AxiosInstance => {
	const instance = axios.create();

	axiosRetry(instance, {
		retries: 5,
		retryDelay: axiosRetry.exponentialDelay,
		retryCondition: (error: AxiosError) => {
			if (error.response) {
				return error.response.status >= 500;
			}
			return axiosRetry.isNetworkError(error);
		},
	});

	return instance;
};
