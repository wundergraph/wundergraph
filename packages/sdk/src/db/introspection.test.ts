import { assert } from 'chai';
import {
	cleanupPrismaSchema,
	introspectPrismaDatabaseWithRetries,
	PrismaDatabaseIntrospectionResult,
} from './introspection';
import fs from 'fs';
import path from 'path';
import { DatabaseIntrospection } from '../definition';

test.skip('introspection', async () => {
	const result = await introspectPrismaDatabaseWithRetries(
		{
			databaseURL: 'postgresql://admin:admin@localhost:54322/example?schema=public',
		},
		'postgresql',
		1
	);
	assert.equal(result.success, true);
});

test('cleanupPrismaSchema', async () => {
	const rawSchema = fs.readFileSync(path.resolve(__dirname, './testdata/schema.graphql.file'));
	const schema = rawSchema.toString('utf-8');
	const introspection: DatabaseIntrospection = {
		databaseURL: 'postgresql://admin:admin@localhost:54322/example?schema=public',
		schemaExtension: `
				type Contact {
					type: String
					phone: String
				}
				input ContactInput {
					type: String
					phone: String
				}
				`,
		replaceCustomScalarTypeFields: [
			{
				entityName: `users`,
				fieldName: `contact`,
				responseTypeReplacement: `Contact`,
				inputTypeReplacement: `ContactInput`,
			},
		],
	};
	const result: PrismaDatabaseIntrospectionResult = {
		graphql_schema: schema,
		interpolateVariableDefinitionAsJSON: [],
		jsonResponseFields: [],
		jsonTypeFields: [],
		message: '',
		prisma_schema: '',
		success: false,
	};

	const cleanedSchema = cleanupPrismaSchema(introspection, result);

	assert.equal(result.jsonTypeFields.length, 1);
	assert.equal(result.jsonResponseFields.length, 0);
	assert.equal(result.interpolateVariableDefinitionAsJSON.length, 2);

	expect(cleanedSchema).toMatchSnapshot();
});
