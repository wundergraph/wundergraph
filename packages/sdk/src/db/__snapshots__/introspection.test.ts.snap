// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanupPrismaSchema 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

input NestedIntFilter {
  equals: Int
  in: [Int]
  notIn: [Int]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input IntFilter {
  equals: Int
  in: [Int]
  notIn: [Int]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

enum QueryMode {
  default
  insensitive
}

input NestedStringFilter {
  equals: String
  in: [String]
  notIn: [String]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input StringFilter {
  equals: String
  in: [String]
  notIn: [String]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  mode: QueryMode
  not: NestedStringFilter
}

enum JsonNullValueFilter {
  DbNull
  JsonNull
  AnyNull
}

input JsonFilter {
  equals: JsonNullValueFilter
  not: JsonNullValueFilter
}

input NestedDateTimeFilter {
  equals: DateTime
  in: [DateTime]
  notIn: [DateTime]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeFilter
}

input DateTimeFilter {
  equals: DateTime
  in: [DateTime]
  notIn: [DateTime]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeFilter
}

input MessagesListRelationFilter {
  every: messagesWhereInput
  some: messagesWhereInput
  none: messagesWhereInput
}

input usersWhereInput {
  AND: usersWhereInput
  OR: [usersWhereInput]
  NOT: usersWhereInput
  id: IntFilter
  email: StringFilter
  name: StringFilter
  contact: JsonFilter
  updatedat: DateTimeFilter
  lastlogin: DateTimeFilter
  messages: MessagesListRelationFilter
}

input UsersRelationFilter {
  is: usersWhereInput
  isNot: usersWhereInput
}

input messagesWhereInput {
  AND: messagesWhereInput
  OR: [messagesWhereInput]
  NOT: messagesWhereInput
  id: IntFilter
  user_id: IntFilter
  message: StringFilter
  users: UsersRelationFilter
}

enum SortOrder {
  asc
  desc
}

input messagesOrderByRelationAggregateInput {
  _count: SortOrder
}

input usersOrderByWithRelationInput {
  id: SortOrder
  email: SortOrder
  name: SortOrder
  contact: SortOrder
  updatedat: SortOrder
  lastlogin: SortOrder
  messages: messagesOrderByRelationAggregateInput
}

input messagesOrderByWithRelationInput {
  id: SortOrder
  user_id: SortOrder
  message: SortOrder
  users: usersOrderByWithRelationInput
}

input messagesWhereUniqueInput {
  id: Int
}

enum MessagesScalarFieldEnum {
  id
  user_id
  message
}

type UsersCountOutputType {
  messages: Int!
}

type users {
  id: Int!
  email: String!
  name: String!
  contact: Contact!
  updatedat: DateTime!
  lastlogin: DateTime!
  messages(where: messagesWhereInput, orderBy: [messagesOrderByWithRelationInput], cursor: messagesWhereUniqueInput, take: Int, skip: Int, distinct: [MessagesScalarFieldEnum]): [messages]
  _count: UsersCountOutputType!
}

type messages {
  id: Int!
  user_id: Int!
  message: String!
  users: users!
}

type MessagesCountAggregateOutputType {
  id: Int!
  user_id: Int!
  message: Int!
  _all: Int!
}

type MessagesAvgAggregateOutputType {
  id: Float
  user_id: Float
}

type MessagesSumAggregateOutputType {
  id: Int
  user_id: Int
}

type MessagesMinAggregateOutputType {
  id: Int
  user_id: Int
  message: String
}

type MessagesMaxAggregateOutputType {
  id: Int
  user_id: Int
  message: String
}

type AggregateMessages {
  _count: MessagesCountAggregateOutputType
  _avg: MessagesAvgAggregateOutputType
  _sum: MessagesSumAggregateOutputType
  _min: MessagesMinAggregateOutputType
  _max: MessagesMaxAggregateOutputType
}

input messagesCountOrderByAggregateInput {
  id: SortOrder
  user_id: SortOrder
  message: SortOrder
}

input messagesAvgOrderByAggregateInput {
  id: SortOrder
  user_id: SortOrder
}

input messagesMaxOrderByAggregateInput {
  id: SortOrder
  user_id: SortOrder
  message: SortOrder
}

input messagesMinOrderByAggregateInput {
  id: SortOrder
  user_id: SortOrder
  message: SortOrder
}

input messagesSumOrderByAggregateInput {
  id: SortOrder
  user_id: SortOrder
}

input messagesOrderByWithAggregationInput {
  id: SortOrder
  user_id: SortOrder
  message: SortOrder
  _count: messagesCountOrderByAggregateInput
  _avg: messagesAvgOrderByAggregateInput
  _max: messagesMaxOrderByAggregateInput
  _min: messagesMinOrderByAggregateInput
  _sum: messagesSumOrderByAggregateInput
}

input NestedFloatFilter {
  equals: Float
  in: [Float]
  notIn: [Float]
  lt: Float
  lte: Float
  gt: Float
  gte: Float
  not: NestedFloatFilter
}

input NestedIntWithAggregatesFilter {
  equals: Int
  in: [Int]
  notIn: [Int]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntWithAggregatesFilter
  _count: NestedIntFilter
  _avg: NestedFloatFilter
  _sum: NestedIntFilter
  _min: NestedIntFilter
  _max: NestedIntFilter
}

input IntWithAggregatesFilter {
  equals: Int
  in: [Int]
  notIn: [Int]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntWithAggregatesFilter
  _count: NestedIntFilter
  _avg: NestedFloatFilter
  _sum: NestedIntFilter
  _min: NestedIntFilter
  _max: NestedIntFilter
}

input NestedStringWithAggregatesFilter {
  equals: String
  in: [String]
  notIn: [String]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringWithAggregatesFilter
  _count: NestedIntFilter
  _min: NestedStringFilter
  _max: NestedStringFilter
}

input StringWithAggregatesFilter {
  equals: String
  in: [String]
  notIn: [String]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  mode: QueryMode
  not: NestedStringWithAggregatesFilter
  _count: NestedIntFilter
  _min: NestedStringFilter
  _max: NestedStringFilter
}

input messagesScalarWhereWithAggregatesInput {
  AND: messagesScalarWhereWithAggregatesInput
  OR: [messagesScalarWhereWithAggregatesInput]
  NOT: messagesScalarWhereWithAggregatesInput
  id: IntWithAggregatesFilter
  user_id: IntWithAggregatesFilter
  message: StringWithAggregatesFilter
}

type MessagesGroupByOutputType {
  id: Int!
  user_id: Int!
  message: String!
  _count: MessagesCountAggregateOutputType
  _avg: MessagesAvgAggregateOutputType
  _sum: MessagesSumAggregateOutputType
  _min: MessagesMinAggregateOutputType
  _max: MessagesMaxAggregateOutputType
}

input usersWhereUniqueInput {
  id: Int
  email: String
}

enum UsersScalarFieldEnum {
  id
  email
  name
  contact
  updatedat
  lastlogin
}

type UsersCountAggregateOutputType {
  id: Int!
  email: Int!
  name: Int!
  contact: Int!
  updatedat: Int!
  lastlogin: Int!
  _all: Int!
}

type UsersAvgAggregateOutputType {
  id: Float
}

type UsersSumAggregateOutputType {
  id: Int
}

type UsersMinAggregateOutputType {
  id: Int
  email: String
  name: String
  updatedat: DateTime
  lastlogin: DateTime
}

type UsersMaxAggregateOutputType {
  id: Int
  email: String
  name: String
  updatedat: DateTime
  lastlogin: DateTime
}

type AggregateUsers {
  _count: UsersCountAggregateOutputType
  _avg: UsersAvgAggregateOutputType
  _sum: UsersSumAggregateOutputType
  _min: UsersMinAggregateOutputType
  _max: UsersMaxAggregateOutputType
}

input usersCountOrderByAggregateInput {
  id: SortOrder
  email: SortOrder
  name: SortOrder
  contact: SortOrder
  updatedat: SortOrder
  lastlogin: SortOrder
}

input usersAvgOrderByAggregateInput {
  id: SortOrder
}

input usersMaxOrderByAggregateInput {
  id: SortOrder
  email: SortOrder
  name: SortOrder
  updatedat: SortOrder
  lastlogin: SortOrder
}

input usersMinOrderByAggregateInput {
  id: SortOrder
  email: SortOrder
  name: SortOrder
  updatedat: SortOrder
  lastlogin: SortOrder
}

input usersSumOrderByAggregateInput {
  id: SortOrder
}

input usersOrderByWithAggregationInput {
  id: SortOrder
  email: SortOrder
  name: SortOrder
  contact: SortOrder
  updatedat: SortOrder
  lastlogin: SortOrder
  _count: usersCountOrderByAggregateInput
  _avg: usersAvgOrderByAggregateInput
  _max: usersMaxOrderByAggregateInput
  _min: usersMinOrderByAggregateInput
  _sum: usersSumOrderByAggregateInput
}

input NestedJsonFilter {
  equals: JsonNullValueFilter
  not: JsonNullValueFilter
}

input JsonWithAggregatesFilter {
  equals: JsonNullValueFilter
  not: JsonNullValueFilter
  _count: NestedIntFilter
  _min: NestedJsonFilter
  _max: NestedJsonFilter
}

input NestedDateTimeWithAggregatesFilter {
  equals: DateTime
  in: [DateTime]
  notIn: [DateTime]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeWithAggregatesFilter
  _count: NestedIntFilter
  _min: NestedDateTimeFilter
  _max: NestedDateTimeFilter
}

input DateTimeWithAggregatesFilter {
  equals: DateTime
  in: [DateTime]
  notIn: [DateTime]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeWithAggregatesFilter
  _count: NestedIntFilter
  _min: NestedDateTimeFilter
  _max: NestedDateTimeFilter
}

input usersScalarWhereWithAggregatesInput {
  AND: usersScalarWhereWithAggregatesInput
  OR: [usersScalarWhereWithAggregatesInput]
  NOT: usersScalarWhereWithAggregatesInput
  id: IntWithAggregatesFilter
  email: StringWithAggregatesFilter
  name: StringWithAggregatesFilter
  contact: JsonWithAggregatesFilter
  updatedat: DateTimeWithAggregatesFilter
  lastlogin: DateTimeWithAggregatesFilter
}

type UsersGroupByOutputType {
  id: Int!
  email: String!
  name: String!
  contact: JSON!
  updatedat: DateTime!
  lastlogin: DateTime!
  _count: UsersCountAggregateOutputType
  _avg: UsersAvgAggregateOutputType
  _sum: UsersSumAggregateOutputType
  _min: UsersMinAggregateOutputType
  _max: UsersMaxAggregateOutputType
}

type Query {
  findFirstmessages(where: messagesWhereInput, orderBy: [messagesOrderByWithRelationInput], cursor: messagesWhereUniqueInput, take: Int, skip: Int, distinct: [MessagesScalarFieldEnum]): messages
  findManymessages(where: messagesWhereInput, orderBy: [messagesOrderByWithRelationInput], cursor: messagesWhereUniqueInput, take: Int, skip: Int, distinct: [MessagesScalarFieldEnum]): [messages]!
  aggregatemessages(where: messagesWhereInput, orderBy: [messagesOrderByWithRelationInput], cursor: messagesWhereUniqueInput, take: Int, skip: Int): AggregateMessages!
  groupBymessages(where: messagesWhereInput, orderBy: [messagesOrderByWithAggregationInput], by: [MessagesScalarFieldEnum]!, having: messagesScalarWhereWithAggregatesInput, take: Int, skip: Int): [MessagesGroupByOutputType]!
  findUniquemessages(where: messagesWhereUniqueInput!): messages
  findFirstusers(where: usersWhereInput, orderBy: [usersOrderByWithRelationInput], cursor: usersWhereUniqueInput, take: Int, skip: Int, distinct: [UsersScalarFieldEnum]): users
  findManyusers(where: usersWhereInput, orderBy: [usersOrderByWithRelationInput], cursor: usersWhereUniqueInput, take: Int, skip: Int, distinct: [UsersScalarFieldEnum]): [users]!
  aggregateusers(where: usersWhereInput, orderBy: [usersOrderByWithRelationInput], cursor: usersWhereUniqueInput, take: Int, skip: Int): AggregateUsers!
  groupByusers(where: usersWhereInput, orderBy: [usersOrderByWithAggregationInput], by: [UsersScalarFieldEnum]!, having: usersScalarWhereWithAggregatesInput, take: Int, skip: Int): [UsersGroupByOutputType]!
  findUniqueusers(where: usersWhereUniqueInput!): users
}

input usersCreateWithoutMessagesInput {
  email: String!
  name: String!
  contact: ContactInput!
  updatedat: DateTime
  lastlogin: DateTime
}

input usersCreateOrConnectWithoutMessagesInput {
  where: usersWhereUniqueInput!
  create: usersCreateWithoutMessagesInput!
}

input usersCreateNestedOneWithoutMessagesInput {
  create: usersCreateWithoutMessagesInput
  connectOrCreate: usersCreateOrConnectWithoutMessagesInput
  connect: usersWhereUniqueInput
}

input messagesCreateInput {
  message: String!
  users: usersCreateNestedOneWithoutMessagesInput!
}

input StringFieldUpdateOperationsInput {
  set: String
}

input DateTimeFieldUpdateOperationsInput {
  set: DateTime
}

input usersUpdateWithoutMessagesInput {
  email: StringFieldUpdateOperationsInput
  name: StringFieldUpdateOperationsInput
  contact: ContactInput
  updatedat: DateTimeFieldUpdateOperationsInput
  lastlogin: DateTimeFieldUpdateOperationsInput
}

input usersUpsertWithoutMessagesInput {
  update: usersUpdateWithoutMessagesInput!
  create: usersCreateWithoutMessagesInput!
}

input usersUpdateOneRequiredWithoutMessagesInput {
  create: usersCreateWithoutMessagesInput
  connectOrCreate: usersCreateOrConnectWithoutMessagesInput
  upsert: usersUpsertWithoutMessagesInput
  connect: usersWhereUniqueInput
  update: usersUpdateWithoutMessagesInput
}

input messagesUpdateInput {
  message: StringFieldUpdateOperationsInput
  users: usersUpdateOneRequiredWithoutMessagesInput
}

input messagesCreateManyInput {
  id: Int
  user_id: Int!
  message: String!
}

type AffectedRowsOutput {
  count: Int!
}

input messagesUpdateManyMutationInput {
  message: StringFieldUpdateOperationsInput
}

input messagesCreateWithoutUsersInput {
  message: String!
}

input messagesCreateOrConnectWithoutUsersInput {
  where: messagesWhereUniqueInput!
  create: messagesCreateWithoutUsersInput!
}

input messagesCreateManyUsersInput {
  id: Int
  message: String!
}

input messagesCreateManyUsersInputEnvelope {
  data: [messagesCreateManyUsersInput]!
  skipDuplicates: Boolean
}

input messagesCreateNestedManyWithoutUsersInput {
  create: messagesCreateWithoutUsersInput
  connectOrCreate: messagesCreateOrConnectWithoutUsersInput
  createMany: messagesCreateManyUsersInputEnvelope
  connect: messagesWhereUniqueInput
}

input usersCreateInput {
  email: String!
  name: String!
  contact: ContactInput!
  updatedat: DateTime
  lastlogin: DateTime
  messages: messagesCreateNestedManyWithoutUsersInput
}

input messagesUpdateWithoutUsersInput {
  message: StringFieldUpdateOperationsInput
}

input messagesUpsertWithWhereUniqueWithoutUsersInput {
  where: messagesWhereUniqueInput!
  update: messagesUpdateWithoutUsersInput!
  create: messagesCreateWithoutUsersInput!
}

input messagesUpdateWithWhereUniqueWithoutUsersInput {
  where: messagesWhereUniqueInput!
  data: messagesUpdateWithoutUsersInput!
}

input messagesScalarWhereInput {
  AND: messagesScalarWhereInput
  OR: [messagesScalarWhereInput]
  NOT: messagesScalarWhereInput
  id: IntFilter
  user_id: IntFilter
  message: StringFilter
}

input messagesUpdateManyWithWhereWithoutUsersInput {
  where: messagesScalarWhereInput!
  data: messagesUpdateManyMutationInput!
}

input messagesUpdateManyWithoutUsersInput {
  create: messagesCreateWithoutUsersInput
  connectOrCreate: messagesCreateOrConnectWithoutUsersInput
  upsert: messagesUpsertWithWhereUniqueWithoutUsersInput
  createMany: messagesCreateManyUsersInputEnvelope
  set: messagesWhereUniqueInput
  disconnect: messagesWhereUniqueInput
  delete: messagesWhereUniqueInput
  connect: messagesWhereUniqueInput
  update: messagesUpdateWithWhereUniqueWithoutUsersInput
  updateMany: messagesUpdateManyWithWhereWithoutUsersInput
  deleteMany: messagesScalarWhereInput
}

input usersUpdateInput {
  email: StringFieldUpdateOperationsInput
  name: StringFieldUpdateOperationsInput
  contact: ContactInput
  updatedat: DateTimeFieldUpdateOperationsInput
  lastlogin: DateTimeFieldUpdateOperationsInput
  messages: messagesUpdateManyWithoutUsersInput
}

input usersCreateManyInput {
  id: Int
  email: String!
  name: String!
  contact: ContactInput!
  updatedat: DateTime
  lastlogin: DateTime
}

input usersUpdateManyMutationInput {
  email: StringFieldUpdateOperationsInput
  name: StringFieldUpdateOperationsInput
  contact: ContactInput
  updatedat: DateTimeFieldUpdateOperationsInput
  lastlogin: DateTimeFieldUpdateOperationsInput
}

type Mutation {
  createOnemessages(data: messagesCreateInput!): messages
  upsertOnemessages(where: messagesWhereUniqueInput!, create: messagesCreateInput!, update: messagesUpdateInput!): messages
  createManymessages(data: [messagesCreateManyInput]!, skipDuplicates: Boolean): AffectedRowsOutput
  deleteOnemessages(where: messagesWhereUniqueInput!): messages
  updateOnemessages(data: messagesUpdateInput!, where: messagesWhereUniqueInput!): messages
  updateManymessages(data: messagesUpdateManyMutationInput!, where: messagesWhereInput): AffectedRowsOutput
  deleteManymessages(where: messagesWhereInput): AffectedRowsOutput
  createOneusers(data: usersCreateInput!): users
  upsertOneusers(where: usersWhereUniqueInput!, create: usersCreateInput!, update: usersUpdateInput!): users
  createManyusers(data: [usersCreateManyInput]!, skipDuplicates: Boolean): AffectedRowsOutput
  deleteOneusers(where: usersWhereUniqueInput!): users
  updateOneusers(data: usersUpdateInput!, where: usersWhereUniqueInput!): users
  updateManyusers(data: usersUpdateManyMutationInput!, where: usersWhereInput): AffectedRowsOutput
  deleteManyusers(where: usersWhereInput): AffectedRowsOutput
}

scalar DateTime

scalar JSON

scalar UUID

scalar BigInt

scalar Bytes

type Contact {
  type: String
  phone: String
}

input ContactInput {
  type: String
  phone: String
}

type Contact {
  type: String
  phone: String
}

input ContactInput {
  type: String
  phone: String
}"
`;
